### PR TITLE
feat(app): add Omi button action toggle

### PR DESCRIPTION
**Pull request by undivisible**
## Summary

- Add an Omi Button Actions toggle to mobile Device Settings, enabled by default.
- Gate Omi BLE button handling and pending voice-command submission on the preference.
- Add localized labels and preference regression coverage.

## Scope

App-only. Firmware is unchanged; firmware-owned behavior such as long-press power-off is not affected.

## Verification

- `bash app/test.sh` — 1,112 tests passed
- `bash app/scripts/analyze_ratchet.sh` — passed
- `flutter test test/unit/omi_button_actions_test.dart` — passed
- Flutter localization generation — stable
- Bounded pre-push checks — passed

<a href="https://cubic.dev/pr/BasedHardware/omi/pull/11193?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

---

> [!NOTE]
> **Medium Risk**
> Changes live BLE button and voice-command submission paths on Omi devices; mistakes could drop legitimate actions or leak audio after the user disables the feature, though cancellation and onboarding guards target that.
> 
> **Overview**
> Adds a user preference **Omi Button Actions** (default **on**) so the mobile app can ignore single/double-tap and device-button voice commands on **Omi** hardware without changing firmware.
> 
> **Device Settings** shows an **Omi Button Actions** switch for `DeviceType.omi`; the **Double Tap** action picker is nested under it and hidden when actions are off. Turning the switch off persists the preference and calls **`cancelActiveVoiceSession()`** so buffered command audio is not sent later.
> 
> **Capture** gates `_handleButtonEvent` and voice submission on the preference for Omi-typed recording devices only; **interactive onboarding** still receives button events, with tighter rules so tutorial voice sessions are cancelled on wrapper teardown and cannot submit after onboarding ends. **`cancelActiveVoiceSession`** plus a submission generation counter also blocks in-flight codec/submit races. **OpenGlass** devices normalized after `getDeviceInfo()` are not subject to this Omi-only gate (voice can still process when actions are “disabled” for Omi).
> 
> **DeviceProvider** re-pushes the capability-normalized paired device into capture after connect so the gate matches post-discovery identity (e.g. Omi advertising vs OpenGlass after image stream).
> 
> Localized label **`omiButtonActions`** and unit/controller tests cover persistence and the gating edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7b26f8684c1d45737bc395415e8f263e5ee5600d. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

Failure-Class: none

## Line-Count-Exception

Line-Count-Exception: app/lib/services/capture/capture_controller.dart | 2900 -> 3116 | Button-actions speaker-haptic injection seam and capability-normalized recording-device updates ride the existing capture lifecycle; file is frozen oversized at 2900 by #9838 and the PR re-applies branch semantics inside main's C1/C2 structure rather than splitting mid-feature.

### CODE REVIEW

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for this @undivisible — clean, well-scoped app feature, and the default-on preference is the right call for preserving current behavior. The l10n coverage across all locales and the hermetic unit test are appreciated.

One blocking issue before this lands:

**Onboarding stranding when the toggle is off.** In `capture_controller.dart` the new guard at line 768 (`if (... !omiButtonActionsEnabled) return;`) runs *before* the interactive-device-onboarding intercept (lines 771–779). That means a user who has turned Omi Button Actions off cannot advance the single/double-press onboarding steps, because `deviceOnboardingProvider.onButtonEvent` never fires. Onboarding button handling isn't really an "action" this toggle should disable. The fix is small: move the gate to just after the onboarding block so onboarding events still flow regardless of the preference.

A couple of smaller follow-ups (non-blocking):
- The `DeviceType.omi && !omiButtonActionsEnabled` check is duplicated verbatim at lines 710 and 768; a small private getter (e.g. `bool get _omiButtonActionsDisabled => ...`) would keep future gate changes in one place.
- The new Switch uses `activeTrackColor: Color(0xFF8E8E93)` (the iOS off-state grey), so it reads as disabled while ON and is inconsistent with the sibling auto-sync toggle (`0xFF8B5CF6`). Matching the existing accent would make the enabled state clearer.
- When the toggle is off, the Double Tap configuration row below it stays visible and tappable even though it's now a no-op — consider hiding/disabling dependent rows, or adding a subtitle.

Notes: CI shows two cancelled `PR Metadata Preflight` runs alongside a passing one; please make sure checks are green before merge. I've added `feature-fit-review` since this is a new user-facing settings surface that warrants a human product/maintainer look, and `needs-maintainer-review` for the onboarding fix. No security/privacy/dependency concerns from my read.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
Omi button action toggle reads clean and is fully localized — feature, so leaving the merge call to Nik.

**Comment by Git-on-my-level** [comment]
Thanks @undivisible — clean, well-scoped implementation. Notes for the human sign-off:

**Implementation**
- Both capture-controller guards (`_processVoiceCommandBytes` and the BLE button stream callback) read the preference at event time, so live toggling works without a reconnect/restart. Default-on preserves current behavior.
- Scope correctly limited to `DeviceType.omi`; other device types are unaffected. Firmware-owned behavior (long-press power-off) is untouched, as stated.

**Minor edge case (non-blocking)**
- If a voice-command session is already active when the user flips the switch off, in-flight bytes are dropped silently and the session lingers until its timeout fires. Not a correctness bug, but dismissing the active session on toggle-off would be slightly cleaner UX. Your call.

**Coverage / CI**
- Unit test covers default value + persistence. `Dart Analyze & Tests`, `Android Compile Smoke`, `Formatting`, `Hygiene`, and `Backend Hermetic Merge Gate` are green on the latest head. The two "PR Metadata Preflight" entries in the check summary are cancelled runs from earlier pushes; the latest run succeeded.

Leaving the product/feature-fit sign-off to a human maintainer.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by Git-on-my-level** [comment]
Follow-up verification on head `0fa4b8c4fd` — all four items from my earlier review are addressed:

1. **Onboarding stranding (was blocking)** — the `_omiButtonActionsDisabled` gate in the BLE button stream now runs *after* the interactive-device-onboarding intercept, so onboarding `onButtonEvent` still fires regardless of the toggle. The same guard in `_processVoiceCommandBytes` remains at the top, which is correct since voice-command bytes are a genuine "action," not onboarding. ✓
2. **Duplicated guard** — consolidated into the `bool get _omiButtonActionsDisabled` private getter; both call sites read through it. ✓
3. **Switch accent** — now `activeTrackColor: Color(0xFF8B5CF6)`, matching the sibling auto-sync toggle. ✓
4. **Dependent Double Tap row** — now hidden when the toggle is off (conditional spread on `_omiButtonActionsEnabled`), with a clean non-Omi fallback path. ✓

Implementation notes:
- Preference defaults to `true`, preserving current behavior.
- Gate is correctly scoped to `DeviceType.omi`; other device types are unaffected.
- Default-on + event-time reads mean live toggling works without reconnect/restart.
- `Dart Analyze & Tests`, `Android Compile Smoke`, `Formatting`, `Generated Files`, and `Backend Hermetic Merge Gate` are green on this head (the earlier `PR Metadata Preflight`/`Hygiene` failures were on the superseded force-pushed commit and are now passing). I dismissed my stale changes-requested review accordingly.

Minor, non-blocking: if a voice-command session is already active when the user flips the toggle off, in-flight bytes are dropped silently and the session lingers until its 15s timeout. Not a correctness issue, but dismissing the active session on toggle-off would be marginally cleaner.

Leaving the product/feature-fit sign-off to a human maintainer — the blocking `feature-fit-review` / `needs-maintainer-review` labels remain.

---
*by AI on behalf of David — if you need David's attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
Omi button action toggle — feature, approve only.

**Review by Git-on-my-level — COMMENTED** [review]
Thanks @undivisible — this is a clean, well-scoped feature. Observations from the diff:

**capture_controller.dart**
- The `_omiButtonActionsDisabled` gate in `streamButton` is correctly placed *after* the onboarding interception, so interactive onboarding still receives button events even when actions are disabled. Good separation of concerns.
- `_isOmiButtonActionsDevice` properly excludes OmiGlass units that advertise as `DeviceType.omi` by name-matching 'glass'/'openglass'/'omiglass', avoiding accidental suppression of OpenGlass behavior.
- `cancelActiveVoiceSession()` correctly cancels the timeout timer, clears `_voiceCommandSession`/`_commandBytes`, and is wired from `device_settings.dart` through `CaptureProvider`, so in-flight audio is dropped rather than submitted after the toggle flips.
- `_processVoiceCommandBytes` early-returns when disabled — consistent with the `streamButton` gate.

**device_settings.dart**
- The customization section now branches on `device?.type == DeviceType.omi`, preserving the existing double-tap row for non-Omi devices. Default-on switch preserves current user behavior.

**preferences.dart**
- Standard `getBool`/`saveBool` pair, default `true`. No migration concern.

**Tests / i18n**
- `omi_button_actions_test.dart` covers default-on and persistence. The capture-gating logic (OmiGlass exclusion, onboarding bypass) isn't unit-tested, which is reasonable given its BLE/device-state dependencies.
- `@omiButtonActions` ARB metadata is present in the English source; generated localization files are consistent (Generated Files check passed).

The `feature-fit-review` and `needs-maintainer-review` labels are retained since the decision to expose this toggle (and its default/naming) is a product call. Note that @kodjima33's prior approvals were on earlier commits (d6572d6, 0fa4b8c); the current head (74733b8b) has moved past them, so a maintainer re-confirmation on this head would close the loop before merge.

Reviewed by glm-5.2 (automation). Leaving for human maintainer review on the current head.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by Git-on-my-level — COMMENTED** [review]
Thanks @undivisible — followed up on this head (`30c37219`) since the prior pass was on `74733b8b`. The review-fix commits since then address the right things:

**capture_controller.dart**
- The generation-based cancellation (`_voiceCommandSubmissionGeneration` captured before the async codec lookup and re-checked after) cleanly prevents stale voice-command audio from being submitted after the user disables actions mid-flight. The test `disabling actions during codec lookup prevents voice submission` exercises exactly this race.
- `_voiceCommandStartedDuringOnboarding` allowing an onboarding-started session to complete even after disable is sensible — the onboarding flow shouldn't be left hanging.

**device_provider.dart**
- `normalizedDevice = pairedDevice ?? device` is a good correctness fix — it ensures the type-resolved paired device (not the raw discovery-type device, which can misclassify OmiGlass as `DeviceType.omi`) reaches `updateRecordingDevice` / `streamDeviceRecording`. This directly supports the `_isOmiButtonActionsDevice` heuristic in capture_controller.

**device_settings.dart**
- Wiring `cancelActiveVoiceSession()` into the toggle's `onChanged` when turning off is the right UX — audio captured while enabled is dropped rather than submitted post-disable.
- The toggle rendering only for `DeviceType.omi` and nesting double-tap config under the enabled state is clean.

All checks green, generated-files check stable, test coverage (5 controller + 2 unit) hits the key invariants. The diff is large but that's almost entirely localization across 40+ locales for a single new key — the logic surface is focused.

Leaving for human maintainer review — the product/UX direction of making device-button actions user-disableable is a behavior choice that warrants maintainer sign-off before merge.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
Omi button action toggle — feature (not bug fix), mobile/app area; approve-only per policy, re-approved on the new head (fix commit landed after last review). Nik/maintainer decides the merge.

**Comment by Git-on-my-level** [comment]
Thanks @undivisible — reviewed the rebase onto `7e91088a1c` (new base `7ebe8b1df2`). The follow-up commits resolve the remaining threads cleanly, and the device-identity handling is now stronger than on the earlier heads:

**capture_controller.dart**
- `_isOmiButtonActionsDevice` is now type-only (`device?.type == DeviceType.omi`) instead of name-matching — sound because `device_provider.dart` pushes the capability-normalized `pairedDevice` into `updateRecordingDevice`, and `_getDeviceInfoFromOmi` reclassifies `hasImageStream` units to `DeviceType.openglass` after connect. The `capability-normalized OpenGlass identity bypasses the Omi-only gate` test pins this.
- `_endVoiceCommandSession` now expires the onboarding exemption (`_voiceCommandStartedDuringOnboarding && deviceOnboardingProvider?.isOnboardingActive == true`), so tutorial-started audio can never be submitted after onboarding exits. Pairs with the wrapper-side `cancelActiveVoiceSession()` in `interactive_device_onboarding_wrapper.dart`, where skip/complete/dispose all funnel through teardown.
- The `collectCommandAudio` gate stops buffering command frames while actions are disabled, closing the last path where a session missed by `cancelActiveVoiceSession()` could keep accumulating audio.
- Legacy long-press (states 3/5) routes through the same `_handleButtonEvent` gate — nothing bypasses the toggle.

**device_provider.dart**
- `normalizedDevice = pairedDevice ?? device` is used for both `updateRecordingDevice` and `streamDeviceRecording`, with a post-`getDeviceInfo()` re-push. This is the load-bearing normalization for the gate; the comments explain the reclassification window well.

**device_settings.dart**
- The switch is Omi-only (`device?.type == DeviceType.omi`), persists immediately, and cancels any in-flight voice session on disable; the Double Tap row nests under the toggle for Omi devices and stays unchanged for others. `context.read<CaptureProvider>()` is safe here since the provider is created in `main.dart` above the navigator.

**Tests & l10n**
- `omi_button_actions_controller_test.dart` covers the races that matter: disable-during-codec-lookup (generation check), onboarding-exit cancellation, stale exemption, and normalized identity. `omiButtonActions` is present in all 49 locale files and the generated localizations are in sync. CI is green on this exact head (Dart Analyze & Tests, Android Compile Smoke, Generated Files, Formatting, Hygiene).
- One transparency note: I could not execute the Dart suite locally (no Flutter SDK available on this review box); the above is static verification against the head plus this head's green CI.

No blocking issues found. What remains is the product gate rather than code: the last maintainer approval (Aug 24) predates this rebase, so feature-fit sign-off on this head — the default-on behavior and hiding vs. disabling the Double Tap row — still needs a human decision.

---
*Feedback generated by AI on behalf of David Zhang (@Git-on-my-level). Human review needed on this head: feature-fit sign-off for the Omi button-actions toggle, since the prior approval predates the rebase onto `7ebe8b1df2`.*

File: app/lib/pages/settings/device_settings.dart [hunk 1]
```diff
@@ -703,20 +704,47 @@ class _DeviceSettingsState extends State<DeviceSettings> {
...
+              icon: FontAwesomeIcons.handPointer,
+              title: context.l10n.omiButtonActions,
+              showChevron: false,
+              trailing: Switch(
+                value: _omiButtonActionsEnabled,
+                activeThumbColor: Colors.white,
+                activeTrackColor: const Color(0xFF8B5CF6),
+                onChanged: (value) {
+                  setState(() => _omiButtonActionsEnabled = value);
+                  SharedPreferencesUtil().omiButtonActionsEnabled = value;
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/pages/settings/device_settings.dart:774]: Fixed on 7e91088a1c / 8efc67aa71: disabling the switch calls `CaptureProvider.cancelActiveVoiceSession()`, which drops the pending bytes, timer, and session so later re-enable cannot submit audio captured while actions were off.

File: app/lib/pages/settings/device_settings.dart [hunk 2]
```diff
@@ -703,20 +705,52 @@ class _DeviceSettingsState extends State<DeviceSettings> {
...
-            title: context.l10n.doubleTap,
-            chipValue: _getDoubleTapActionLabel(doubleTapAction),
-            onTap: _showDoubleTapActionSheet,
-          ),
+          if (device?.type == DeviceType.omi) ...[
+            _buildProfileStyleItem(
+              icon: FontAwesomeIcons.handPointer,
+              title: context.l10n.omiButtonActions,
+              showChevron: false,
+              trailing: Switch(
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/pages/settings/device_settings.dart:767]: Fixed: the switch now has `Key('omi_button_actions_toggle')` so agent-flutter / E2E can target it without coordinates.

File: app/lib/pages/settings/device_settings.dart [hunk 3]
```diff
@@ -757,13 +759,46 @@ class _DeviceSettingsState extends State<DeviceSettings> {
...
             const Divider(height: 1, color: Color(0xFF3C3C43)),
           ],
-          // Double Tap
-          _buildProfileStyleItem(
-            icon: FontAwesomeIcons.handPointer,
-            title: context.l10n.doubleTap,
-            chipValue: _getDoubleTapActionLabel(doubleTapAction),
-            onTap: _showDoubleTapActionSheet,
-          ),
+          if (device?.type == DeviceType.omi) ...[
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/pages/settings/device_settings.dart:762]: Fixed in 8efc67aa71: customization (including this Omi-only toggle) now uses `pairedDevice ?? connectedDevice`, so OpenGlass reclassified after `hasImageStream` is not shown a switch that its actions bypass.

File: app/lib/pages/onboarding/interactive_device_onboarding/interactive_device_onboarding_wrapper.dart [hunk 4]
```diff
@@ -60,6 +60,12 @@ class _InteractiveDeviceOnboardingWrapperState extends State<InteractiveDeviceOn
     if (_started && !_completed) {
       AnalyticsManager().deviceOnboardingAbandoned(_onboardingProvider.currentStep);
     }
+    // Cancel any tutorial-started voice-command session (single press on step 1)
+    // before detaching. Skip/complete both pop, so every exit funnels through
+    // here; without this the controller session and its 15s timer outlive the
+    // tutorial and the started-during-onboarding exemption could submit audio
+    // captured after the user left.
+    _captureProvider?.cancelActiveVoiceSession();
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/pages/onboarding/interactive_device_onboarding/interactive_device_onboarding_wrapper.dart]: Fixed in 8efc67aa71: dispose cancels the voice session only after `_startTutorial` (`!_showIntro`). Viewing Settings → Device Tutorial and leaving the intro no longer tears down an unrelated in-flight command.

File: app/lib/providers/device_provider.dart [hunk 5]
```diff
@@ -597,8 +597,9 @@ class DeviceProvider extends ChangeNotifier implements IDeviceServiceSubsciption
       return;
     }
 
+    final normalizedDevice = pairedDevice ?? device;
     if (captureProvider != null) {
-      captureProvider?.updateRecordingDevice(device);
+      captureProvider?.updateRecordingDevice(normalizedDevice);
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/providers/device_provider.dart:618]: Fixed in cbbda025a3. Added `DeviceProvider.capabilityNormalizedDevice` (prefer same-id `pairedDevice`) and wired Home + speech-profile restarts through it. `streamDeviceRecording` / `updateRecordingDevice` no longer downgrade an already-normalized OpenGlass type back to advertising-time Omi.

File: app/lib/pages/onboarding/interactive_device_onboarding/interactive_device_onboarding_wrapper.dart [hunk 6]
```diff
@@ -60,6 +60,13 @@ class _InteractiveDeviceOnboardingWrapperState extends State<InteractiveDeviceOn
     if (_started && !_completed) {
       AnalyticsManager().deviceOnboardingAbandoned(_onboardingProvider.currentStep);
     }
+    // Cancel only a tutorial-owned voice session. Opening Settings → Device
+    // Tutorial attaches CaptureProvider in the post-frame callback before the
+    // user starts the flow; exiting the intro must not discard an unrelated
+    // in-flight Omi voice command. `_showIntro` stays true until `_startTutorial`.
+    if (!_showIntro) {
+      _captureProvider?.cancelActiveVoiceSession();
+    }
```
[the remark this answers is not in the document]
Comment (undivisible) [reply app/lib/pages/onboarding/interactive_device_onboarding/interactive_device_onboarding_wrapper.dart:69]: Fixed in cbbda025a3. Tutorial dispose now calls `cancelTutorialOwnedVoiceSession()`, which no-ops unless `_voiceCommandStartedDuringOnboarding` is set. Tapping Start then exiting on step 0 no longer discards a pre-existing Omi command.

**Comment by Git-on-my-level** [comment]
Thanks @undivisible — reviewed through cbbda025a3 (the rebase plus the capabilityNormalizedDevice and tutorial-cancel follow-ups). The device-identity thread is now handled much more cleanly than the earlier name-matching approach, and the edge-case tests are genuinely good. Remaining items on this head are two mechanical CI fixes plus one behavioral question:

**Verified in the diff**
- capture_controller.dart — _isOmiButtonActionsDevice is now purely type-based (device?.type == DeviceType.omi), sound given DeviceProvider pushes the capability-normalized identity; the _voiceCommandSubmissionGeneration counter re-checked after the codec await plus the collectCommandAudio guard close the disable-mid-session leak from both ends; _endVoiceCommandSession expires the onboarding exemption once onboarding is inactive; cancelTutorialOwnedVoiceSession only touches tutorial-owned sessions.
- device_provider.dart — capabilityNormalizedDevice preferring the same-id pairedDevice, with _recordingDevicePreservingNormalizedType in the capture controller preventing restarts from re-downgrading a normalized OpenGlass.
- interactive_device_onboarding_wrapper.dart — the !_showIntro guard means merely viewing Settings → Device Tutorial no longer discards an unrelated in-flight voice command.
- device_settings.dart — stable Key('omi_button_actions_toggle'), Double Tap nested under the toggle and hidden while disabled, cancelActiveVoiceSession() on disable, pairedDevice ?? connectedDevice for customization. preferences.dart default-on persistence is pinned by omi_button_actions_test.dart.

**Blocking CI on this head (both small)**
1. Formatting — CI flags app/lib/pages/onboarding/speech_profile_widget.dart as needing dart format (exit 123, 1 changed file).
2. Dart Analyze & Tests — the only failing test is the new 'streamDeviceRecording does not downgrade a normalized OpenGlass type', failing with 'Service manager is not initiated' from _closeBleStream. streamDeviceRecording → _resetState → _closeBleStream reaches ServiceManager.instance(), and unlike sibling files (capture_provider_test.dart, device_provider_test.dart), this test file's setUp doesn't bootstrap ServiceManager.init() with the usual try/catch. Looks like a test-bootstrap gap rather than a production issue (main.dart initializes the service manager before any UI), but the PR description's '1,112 tests passed' no longer reflects this head.

(The Backend Hermetic Merge Gate red is just the cancelled scope-detection job; no backend files are touched.)

**One behavioral question on capabilityNormalizedDevice (from the latest Codex pass, worth resolving before merge)**
- After a disconnect, getDeviceInfo() repopulates pairedDevice from saved preferences while connectedDevice stays null (device_provider.dart:231-236), so capabilityNormalizedDevice returns the remembered device. Home entry (pages/home/page.dart:502) now feeds that into streamDeviceRecording, which sets _recordingDevice and — since _getAudioCodec falls back to pcm8 when ensureConnection returns null — can open a transcription socket for a device with no BLE connection. Before this PR, connectedDevice == null meant check-only. If the intent is only to preserve normalized identity for connected devices, gating the getter on connected != null would keep the old disconnected behavior; if opening the socket for the remembered device is intended so reconnect resumes faster, a short comment saying so would help.

None of this changes the positive read on the feature itself. With the two CI fixes and a word on the disconnected-identity question, this should be ready for maintainer sign-off on the current head (the earlier approvals predate these commits).

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Review by kodjima33 — APPROVED** [review]
feat(app): Omi button action toggle — feature, not a bug fix (mobile area), approve-only per policy. Re-confirmed on latest commit.

**Review by kodjima33 — APPROVED** [review]
Feature (Omi button action toggle) — approve-only per policy (mobile area, not a bug fix). Note: CI currently shows failures (Dart Analyze, Formatting, Hermetic Merge Gate) — flagging for visibility, not blocking approval.

**Review by kodjima33 — APPROVED** [review]
Mobile feature (Omi button action toggle) from flagged author, full diff re-read finds no abuse signal (pure app/l10n/provider files); DIRTY with failing required checks (Dart Analyze, Formatting, Backend Hermetic Merge Gate) so stays approve-only under the hard floor regardless of area rules.

**Review by kodjima33 — APPROVED** [review]
Legit feature (Omi button-actions toggle gating BLE button/voice-command handling on preference), diff matches description, no abuse signals found. Mobile feature area -> approve-only per policy (not a bug fix). DIRTY mergeStateStatus + multiple real CI failures (Backend Hermetic Merge Gate, Dart Analyze, Formatting) block merge regardless — author should rebase and fix CI.

**Comment by undivisible** [comment]
Rebased onto current `main` (head `8e149c14`).

**Two mechanical CI fixes pushed:**

1. **Legacy long-press flag resurrection.** The rebase had resurrected `_voiceSessionStartedByLegacyLongPress` from the pre-rebase branch while main removed it in #12477 — 6 undefined references in `capture_controller.dart`. Re-removed; behavior now matches main's `604e2f6e5d` exactly (single-tap always ends a session, button-5 release ends any active session).

2. **Formatter drift** in `speech_profile_widget.dart` (`dart format` line-length wrap) — committed as a style fix.

**Rebase conflict resolutions** (`capture_provider.dart`, `capture_controller.dart`): kept main's `microphonePermissionRequester`/`phoneMicBatchRecorder`/`recordingTelemetry` additions AND this PR's `speakerHaptic` injection; kept main's inline button handling? No — took the PR's `_handleButtonEvent` extraction (testability) and merged it with main's geolocation/WAL additions.

**On the `capabilityNormalizedDevice` disconnected-device question** (from the Aug 25 review): the current shape returns the *remembered* device after a disconnect, and `streamDeviceRecording` falls back to `pcm8` when `ensureConnection` returns null — so yes, a transcription socket could open for an unconnected remembered device. Gating the getter on `connectedDevice != null` would restore the old check-only behavior, but it would also break the reconnect path that reads `streamDeviceRecording` with the normalized identity. The safer minimal fix: gate inside `streamDeviceRecording` — if `ensureConnection` is null and the device is not connected, skip socket setup and log. Happy to push that as a follow-up commit if that matches intent, or leave as-is if opening the socket for the remembered device is intended so reconnect streams immediately.

**Verification:** 10/10 `omi_button_actions_controller_test.dart` + 2/2 `omi_button_actions_test.dart` pass; `flutter analyze` clean on all 5 touched files; `dart format` clean.

**Comment by Git-on-my-level** [comment]
Thanks @undivisible — followed up on the rebased head `8e149c14`. Both items from my last pass are confirmed fixed in the diff and CI:

- **Formatting** — `8e149c1` wraps the `speech_profile_widget.dart` line; the Formatting check is green on this head.
- **Test bootstrap** — `24b46a0` swaps the raw `CaptureProvider` fixtures for `_NoSocketCaptureProvider` (overriding `openConversationSocket` to return null so tests never reach `buildHeaders` → FirebaseAuth), stubs `path_provider`/`package_info` in `setUpAll`, seeds `deviceIdHash`, and boots `PlatformManager.initializeServices()` per test. Dart Analyze & Tests is green.
- The rebase-resurrected `_voiceSessionStartedByLegacyLongPress` flag is fully gone on this head (0 references); single-tap-ends and release-ends semantics match main's post-#12477 shape.

On the `capabilityNormalizedDevice` question from the Aug 25 review — I traced the disconnected-device path on this head and it's a bit sharper than "could open a socket":

- `DeviceProvider.onDeviceDisconnected()` nulls `connectedDevice` but `pairedDevice` survives, so `capabilityNormalizedDevice` returns the remembered device, and the `initState` post-frame callback in `pages/home/page.dart:502` now passes it into `streamDeviceRecording` (pre-PR this was `connectedDevice == null` → check-only).
- `streamDeviceRecording` then runs the device-requested path: `_resetState` → `_ensureDeviceSocketConnection` → `_getAudioCodec` → `ensureConnection` (force=false) returns null for the disconnected device → codec falls back to `pcm8` → `_initiateWebsocket(force: true)` opens a real transcription socket. `_initiateDeviceAudioStreaming` does bail out early (`connection == null` → return), so no audio ever flows — but the socket is open for a device with no BLE link.
- Two side effects ride along: `deviceRequested` is now true, so `_recordingTelemetry.prepare()` mints a recording ID and `failStart(failureClass: 'capture_unavailable')` fires when nothing starts — every cold start with a paired-but-disconnected pendant gets classified as a failed start — and `_captureSessionLocation(promptIfDenied: device != null)` now takes the prompt-if-denied branch on plain app start.

On the fix shape: reconnect doesn't depend on this socket — `_onConnected` → `connectDevice` re-runs `streamDeviceRecording` with the normalized identity and opens the socket anyway — so gating `_ensureDeviceSocketConnection` on `ensureConnection(deviceId) != null` (your proposed minimal gate) looks safe from static reading and restores the old check-only behavior for the disconnected case. What remains is intent rather than mechanics: whether an open-but-idle socket for the remembered pendant is acceptable UX (and whether the backend is fine with the empty `pcm8` session), and whether the cold-start `capture_unavailable` telemetry is the classification you want. That's a maintainer ruling on the current head — noting the earlier approvals all predate the rebase to `8e149c14`.

With that ruling plus the small gate (or a comment stating the behavior is intended), this is ready for fresh maintainer sign-off.

---
*by AI on behalf of David.*

**Review by kodjima33 — APPROVED** [review]
Feature (Omi button action toggle), not a bug fix — approve-only per policy confidence gate. Clean scoped diff (preferences.dart + localization only), no abuse signals. mergeStateStatus=UNKNOWN so holding at approve, not merging.

**Review by kodjima33 — APPROVED** [review]
Feature (Omi button-actions toggle), mobile/app area — approve-only per policy (features stay approve-only outside desktop).

**Review by kodjima33 — APPROVED** [review]
Mobile feature (button-action toggle) — approve-only per policy (not a bug fix); re-affirming approval on backlog re-eval.

**Review by kodjima33 — APPROVED** [review]
feat(app) Omi button action toggle -- feature not bug fix, approve-only; re-approving since reviewDecision reverted to REVIEW_REQUIRED.

**Review by kodjima33 — APPROVED** [review]
feat(app) Omi button action toggle -- near-duplicate of #12305, feature not bug fix, approve-only; re-approving since reviewDecision reverted to REVIEW_REQUIRED again.

**Review by kodjima33 — APPROVED** [review]
Feature (mobile): Omi button actions toggle. Clean, well-tested gating of BLE button/voice-command paths behind a persisted preference; onboarding exemption and race-safe cancellation (submission generation counter) reviewed in full given elevated-scrutiny author. No abuse signals. Approve-only per policy (feature, not bug fix).

**Review by kodjima33 — APPROVED** [review]
app-only feature (Omi button action toggle), gated behind preference; approve per policy (feature in mobile/community area stays approve-only, no merge).

**Comment by Git-on-my-level** [comment]
Merge-forward performed by @dazheng_hermes_macstudio_bot (authorized by David): merged `origin/main` into this branch (`8e149c1..ca2a1a3`, merge commit `ca2a1a3e54`), no rebase, no force-push.

Conflict resolutions (main's C1/C2 capture-session structure kept, branch semantics re-applied):
- `capture_controller.dart`: kept main's session/lifetime constructor + handler structure; re-applied the branch's `speakerHaptic` injection seam (param + `_speakerHaptic` field + short-circuit in `_playSpeakerHaptic`) and type-preserving recording-device updates (`_recordingDevicePreservingNormalizedType` around main's `_rollCaptureSession`). Main's computed `isPaused => _preferences.deviceMuted` supersedes the branch's startup mute-restore (same outcome, main's mechanism).
- `device_provider.dart`: union — branch's `capabilityNormalizedDevice` getter + normalized recording-device pushes kept; main's generation guards kept.
- `speech_profile_widget.dart`: main's AnimatedBuilder rewrite kept; branch's getter swap is subsumed (restart now flows through `capture.recordingDevice`, which the device_provider normalization feeds).
- `capture_provider.dart`: union of constructor params (`super.speakerHaptic` + main's new seams).
- Test stub updated for main's new `geolocation` socket param; `Line-Count-Exception` declared in PR body for `capture_controller.dart` (2900 -> 3116).

Verified locally at `ca2a1a3e54`: `dart analyze lib` 0 errors (142 pre-existing infos); `flutter test` button-actions suites **12/12 pass**; ownership audit clean (net diff = exactly the branch's 110 owned files; 29 main-side files restored to origin/main bytes after formatter drift).
